src/io/network/addrinfo.hpp

```diff
@@ -11,21 +11,39 @@
 
 #pragma once
 
+#include <netdb.h>
+#include <string>
+
 namespace io::network {
 
+struct Endpoint;
+
 /**
  * Wrapper class for getaddrinfo.
  * see: man 3 getaddrinfo
  */
 class AddrInfo {
-  explicit AddrInfo(struct addrinfo *info);
-
  public:
+  struct Iterator {
+    explicit Iterator(addrinfo *p) : ptr(p) {}
+    addrinfo &operator*() const { return *ptr; }
+    Iterator &operator++() {
+      ptr = ptr->ai_next;
+      return *this;
+    }
+    bool operator!=(const Iterator &rhs) { return rhs.ptr != ptr; };
```

Comment:
This break `operator symmetry` meaning that for two iterators `a && b` `a != b` is not the same as `b != a` because  the lhs is subject to different conversion rules than the right (by C++ standard). Define it an `inline friend` :)

Reply:
I couldn't find the exact change in the standard, but the spaceship operator makes "reversed ordering" of comparison operators [possible](https://godbolt.org/z/vhn9Txjsd).

Reply:
Yes but it's still not `symmetrical` unless defined as friend. Simply put the expression 
`a == b` is transformed to `a.operator==(b)` which  is not symmetrical to `b.operartor==(a)`. However if you make it a function then both `a == b` and `b == a` are the exact same thing from the standards point of view

Reply:
cool, I have made the op it a friend function

Reply:
@sinisa-susnjar Something very very interesting! So take a look here: https://godbolt.org/z/qbYsfrz7a 
It turns out that C++ 20 added reversed operators for comparisons so the above code is actually symmetrical (remove the C++20 standard from godbolt and you should get an error). 
p.s.  special thanks for discussing this with @antaljanosbenjamin my mind is bend! 
p.s. 2. My comment still applies for ALL other operators, so keep that in mind!